domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +208,340 @@ class TopicListController @Inject constructor(
+
+    sortedTopicProgressList.forEach { topicProgress ->
       val topic = topicController.retrieveTopic(topicProgress.topicId)
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
         val story = topicController.retrieveStory(topic.topicId, storyId)
 
-        val completedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
         val lastCompletedChapterProgress: ChapterProgress? =
           completedChapterProgressList.firstOrNull()
 
-        val startedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
 
         val recentlyPlayerChapterProgress: ChapterProgress? =
           startedChapterProgressList.firstOrNull()
-        if (recentlyPlayerChapterProgress != null) {
-          val recentlyPlayerChapterSummary: ChapterSummary? =
-            story.chapterList.find { chapterSummary ->
-              recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
-            }
-          if (recentlyPlayerChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
-              storyId,
-              topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              recentlyPlayerChapterSummary.name,
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+
+        when {
+          recentlyPlayerChapterProgress != null -> {
+            val recentlyPlayerChapterSummary: ChapterSummary? =
+              story.chapterList.find { chapterSummary ->
+                recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
+              }
+            if (recentlyPlayerChapterSummary != null) {
+              val numberOfDaysPassed = (
+                oppiaClock.getCurrentCalendar().timeInMillis -
+                  recentlyPlayerChapterProgress.lastPlayedTimestamp
+                ) / ONE_DAY_IN_MS
```

Comment:
Here, above, and below: if it simplifies things, this might benefit from using an extension function:

```kotlin
private fun ChapterProgress.getNumberOfDaysPassed(): Int {
  return TimeUnit.MILLIS.toDays(
    oppiaClock.getCurrentCalendar().timeInMillis - this.lastPlayedTimestamp
  )
}
```

Reply:
Done here and everywhere else.